lib/chef/knife.rb

```diff
@@ -343,39 +343,38 @@ def parse_options(args)
       exit(1)
     end
 
-    # keys from mixlib-cli options
-    def cli_keys
-      self.class.options.keys
+    # This is all set and default mixlib-config values.  We only need the default
+    # values here (the set values are explicitly mixed in again later), but there is
+    # no mixlib-config API to get a Hash back with only the default values.
+    #
+    # Assumption:  since base_mixlib_config is the lowest precedence it doesn't matter
+    # that we include the set values here.
+    #
+    def base_mixlib_config
+      Chef::Config[:knife].save(true)
     end
 
-    # extracts the settings from the Chef::Config[:knife] sub-hash that correspond
-    # to knife cli options -- in preparation for merging config values with cli values
+    # This is only the user-set mixlib-config values.  We do not include the defaults
+    # here so that the config defaults do not override the cli defaults.
     #
-    # NOTE: due to weirdness in mixlib-config #has_key? is only true if the value has
-    # been set by the user -- the Chef::Config defaults return #has_key?() of false and
-    # this code DEPENDS on that functionality since applying the default values in
-    # Chef::Config[:knife] would break the defaults in the cli that we would otherwise
-    # overwrite.
     def config_file_settings
-      cli_keys.each_with_object({}) do |key, memo|
-        if Chef::Config[:knife].key?(key)
-          memo[key] = Chef::Config[:knife][key]
-        end
-      end
+      Chef::Config[:knife].save(false)
     end
 
     # config is merged in this order (inverse of precedence)
+    #  base_mixlib_config   - Chef::Config[:knife] defaults
```

Comment:
This method could really use a more clear name. I get the order after reading it here, but seeing these methods used it's really not apparent where the config data is coming from

Reply:
Yeah I'm open to a better name but I'm also at a loss.  I would have used `default_config` because that's what its supposed to be, but mixlib-cli already took that.  So I went with `default_mixlib_config`, but then its not really just the defaults, although that's the _intent_ but it actually includes all of mixlib-config.  So I went with "base_mixlib_config" because its really the base config that everything is layered on.  I hate it all.